File: js/src/crdt/map.ts
/**
 * LWW-Map (Last-Writer-Wins Map) CRDT implementation.
 *
 * A map where each key is independently an LWW-Register.
 * Supports nested paths for deep object structures.
 */

import {
  BaseCRDT,
  CRDT,
  Operation,
  StateCRDT,
  createOperation,
  operationFromDict,
  operationToDict,
  isDangerousKey,
  validatePath,
  validateValue,
} from "./types.js";

/** Entry value with timestamp metadata */
type Entry = [value: unknown, timestamp: number, nodeId: string];

/** Tombstone with timestamp metadata */
type Tombstone = [timestamp: number, nodeId: string];

/**
 * Last-Writer-Wins Map.
 *
 * Each key in the map is independently resolved using LWW semantics.
 * Supports nested paths like ["users", "123", "name"] for deep updates.
 *
 * @example
 * ```typescript
 * const state = new LWWMap("node-1");
 * state.set(["user", "name"], "Alice");
 * state.set(["user", "age"], 30);
 * console.log(state.value()); // { user: { name: "Alice", age: 30 } }
 * ```
 */
export class LWWMap
  extends BaseCRDT<Record<string, unknown>>
  implements StateCRDT<Record<string, unknown>>
{
  /** Store (value, timestamp, nodeId) for each path */
  private entries: Map<string, Entry> = new Map();

  /** Track deleted paths */
  private tombstones: Map<string, Tombstone> = new Map();

  constructor(
    nodeId: string,
    initialValue?: Record<string, unknown> | null
  ) {
    super(nodeId);
    if (initialValue) {
      this.initFromValue(initialValue);
    }
  }

  /**
   * Convert a path array to a string key for Map storage.
   */
  private pathToKey(path: string[]): string {
    return path.join("\x00"); // Use null character as separator
  }

  /**
   * Convert a string key back to a path array.
   */
  private keyToPath(key: string): string[] {
    return key.split("\x00");
  }

  /**
   * Initialize entries from a nested dictionary.
   */
  private initFromValue(
    value: Record<string, unknown>,
    path: string[] = []
  ): void {
    for (const [key, val] of Object.entries(value)) {
      // Skip dangerous keys to prevent prototype pollution
      if (isDangerousKey(key)) {
        continue;
      }
      const currentPath = [...path, key];
      if (val !== null && typeof val === "object" && !Array.isArray(val)) {
        this.initFromValue(val as Record<string, unknown>, currentPath);
      } else {
        this.entries.set(this.pathToKey(currentPath), [val, 0, this.nodeId]);
      }
    }
  }

  /**
   * Set a value at the given path.
   *
   * If value is an object, it will be flattened into individual path entries.
   * @returns The operation that was applied.
   */
  set(path: string[], value: unknown): Operation {
    const op = createOperation(this.nodeId, path, "set", value);
    this.apply(op);
    return op;
  }

  /**
   * Delete a value at the given path.
   *
   * @returns The operation that was applied.
   */
  delete(path: string[]): Operation {
    const op = createOperation(this.nodeId, path, "delete");
    this.apply(op);
    return op;
  }

  /**
   * Get a value at the given path.
   */
  get(path: string[]): unknown {
    const pathKey = this.pathToKey(path);

    // Check if path is deleted
    const tombstone = this.tombstones.get(pathKey);
    if (tombstone) {
      const entry = this.entries.get(pathKey);
      if (!entry || tombstone[0] > entry[1]) {
        return undefined;
      }
    }

    const entry = this.entries.get(pathKey);
    if (entry) {
      return entry[0];
    }

    // Check for nested values
    return this.getNested(path);
  }

  /**
   * Get all values nested under a path as a dictionary.
   * Filters out dangerous keys to prevent prototype pollution.
   */
  private getNested(path: string[]): Record<string, unknown> | undefined {
    const result: Record<string, unknown> = {};
    const pathKey = this.pathToKey(path);
    const pathLen = path.length;
    let hasValues = false;

    for (const [entryKey, [value, ts]] of this.entries) {
      const entryPath = this.keyToPath(entryKey);

      // Skip any paths containing dangerous keys
      if (entryPath.some(isDangerousKey)) {
        continue;
      }

      if (entryPath.length > pathLen && entryKey.startsWith(pathKey + "\x00")) {
        // Check tombstone
        const tombstone = this.tombstones.get(entryKey);
        if (tombstone && tombstone[0] > ts) {
          continue;
        }

        // Build nested structure safely
        const remaining = entryPath.slice(pathLen);
        let current = result;
        for (let i = 0; i < remaining.length - 1; i++) {
          const key = remaining[i];
          // Safety check for dangerous keys
          if (isDangerousKey(key)) {
            continue;
          }
          if (!(key in current)) {
            current[key] = {};
          }
          current = current[key] as Record<string, unknown>;
        }
        const finalKey = remaining[remaining.length - 1];
        if (!isDangerousKey(finalKey)) {
          current[finalKey] = value;
          hasValues = true;
        }
      }
    }

    return hasValues ? result : undefined;
  }

  /**
   * Apply a set or delete operation.
   */
  apply(op: Operation): boolean {
    if (this.hasSeen(op)) {
      return false;
    }

    const path = op.path;

    if (op.opType === "set") {
      this.applySet(path, op.value, op.timestamp, op.nodeId);
    } else if (op.opType === "delete") {
      this.applyDelete(path, op.timestamp, op.nodeId);
    } else {
      throw new Error(
        `LWWMap supports 'set' and 'delete' operations, got '${op.opType}'`
      );
    }

    this.recordOperation(op);
    return true;
  }

  /**
   * Apply a set operation at a path.
   */
  private applySet(
    path: string[],
    value: unknown,
    timestamp: number,
    nodeId: string
  ): void {
    if (value !== null && typeof value === "object" && !Array.isArray(value)) {
      // Flatten nested object into individual entries
      this.flattenSet(path, value as Record<string, unknown>, timestamp, nodeId);
    } else {
      const pathKey = this.pathToKey(path);
      const existing = this.entries.get(pathKey);
      if (!existing || this.isNewer(timestamp, nodeId, existing[1], existing[2])) {
        this.entries.set(pathKey, [value, timestamp, nodeId]);
      }
    }
  }

  /**
   * Flatten a nested object into individual path entries.
   * Skips dangerous keys to prevent prototype pollution.
   */
  private flattenSet(
    path: string[],
    value: Record<string, unknown>,
    timestamp: number,
    nodeId: string
  ): void {
    for (const [key, val] of Object.entries(value)) {
      // Skip dangerous keys to prevent prototype pollution
      if (isDangerousKey(key)) {
        continue;
      }
      const currentPath = [...path, key];
      if (val !== null && typeof val === "object" && !Array.isArray(val)) {
        this.flattenSet(currentPath, val as Record<string, unknown>, timestamp, nodeId);
      } else {
        const pathKey = this.pathToKey(currentPath);
        const existing = this.entries.get(pathKey);
        if (!existing || this.isNewer(timestamp, nodeId, existing[1], existing[2])) {
          this.entries.set(pathKey, [val, timestamp, nodeId]);
        }
      }
    }
  }

  /**
   * Apply a delete operation at a path.
   */
  private applyDelete(
    path: string[],
    timestamp: number,
    nodeId: string
  ): void {
    const pathKey = this.pathToKey(path);
    const existing = this.tombstones.get(pathKey);
    if (!existing || this.isNewer(timestamp, nodeId, existing[0], existing[1])) {
      this.tombstones.set(pathKey, [timestamp, nodeId]);
    }
  }

  /**
   * Check if (ts1, node1) is newer than (ts2, node2).
   */
  private isNewer(
    ts1: number,
    node1: string,
    ts2: number,
    node2: string
  ): boolean {
    if (ts1 !== ts2) {
      return ts1 > ts2;
    }
    return node1 > node2;
  }

  /**
   * Merge another map into this one.
   */
  merge(other: CRDT<Record<string, unknown>>): void {
    for (const op of other.allOperations()) {
      this.apply(op);
    }
  }

  /**
   * Get the current value as a nested dictionary.
   * Filters out dangerous keys to prevent prototype pollution.
   */
  value(): Record<string, unknown> {
    const result: Record<string, unknown> = {};

    for (const [pathKey, [value, ts]] of this.entries) {
      const path = this.keyToPath(pathKey);

      // Skip any paths containing dangerous keys
      if (path.some(isDangerousKey)) {
        continue;
      }

      // Check tombstone
      const tombstone = this.tombstones.get(pathKey);
      if (tombstone && tombstone[0] > ts) {
        continue;
      }

      // Build nested structure safely
      let current = result;
      for (let i = 0; i < path.length - 1; i++) {
        const key = path[i];
        // Double-check for dangerous keys (defense in depth)
        if (isDangerousKey(key)) {
          continue;
        }
        if (!(key in current)) {
          current[key] = {};
        } else if (typeof current[key] !== "object" || current[key] === null) {
          // Conflict: path has both a value and nested values
          // Nested values take precedence
          current[key] = {};
        }
        current = current[key] as Record<string, unknown>;
      }

      if (path.length > 0) {
        const finalKey = path[path.length - 1];
        // Final safety check
        if (!isDangerousKey(finalKey)) {
          current[finalKey] = value;
        }
      }
    }

    return result;
  }

  /**
   * Get the full state for transmission.
   */
  state(): Record<string, unknown> {
    const entries: Record<string, unknown> = {};
    for (const [pathKey, [val, ts, nid]] of this.entries) {
      const path = this.keyToPath(pathKey);
      entries[path.join(".")] = { value: val, timestamp: ts, node_id: nid };
    }

    const tombstones: Record<string, unknown> = {};
    for (const [pathKey, [ts, nid]] of this.tombstones) {
      const path = this.keyToPath(pathKey);
      tombstones[path.join(".")] = { timestamp: ts, node_id: nid };
    }

    return {
      entries,
      tombstones,
      operations: this.operations.map(operationToDict),
    };
  }

  /**
   * Reconstruct map from transmitted state.
   */
  static fromState(
    nodeId: string,
    state: Record<string, unknown>
  ): LWWMap {
    const lwwMap = new LWWMap(nodeId);

    const entries = state.entries as Record<string, Record<string, unknown>> ?? {};
    for (const [pathStr, entry] of Object.entries(entries)) {
      const path = pathStr ? pathStr.split(".") : [];
      const pathKey = lwwMap.pathToKey(path);
      lwwMap.entries.set(pathKey, [
        entry.value,
        entry.timestamp as number,
        entry.node_id as string,
      ]);
    }

    const tombstones = state.tombstones as Record<string, Record<string, unknown>> ?? {};
    for (const [pathStr, tombstone] of Object.entries(tombstones)) {
      const path = pathStr ? pathStr.split(".") : [];
      const pathKey = lwwMap.pathToKey(path);
      lwwMap.tombstones.set(pathKey, [
        tombstone.timestamp as number,
        tombstone.node_id as string,
      ]);
    }

    lwwMap.operations = (
      (state.operations as Record<string, unknown>[]) ?? []
    ).map(operationFromDict);

    return lwwMap;
  }

  /**
   * Get top-level keys.
   */
  keys(): string[] {
    const seen = new Set<string>();
    for (const pathKey of this.entries.keys()) {
      const path = this.keyToPath(pathKey);
      if (path.length > 0) {
        seen.add(path[0]);
      }
    }
    return Array.from(seen);
  }

  /**
   * Check if a top-level key exists.
   */
  has(key: string): boolean {
    for (const pathKey of this.entries.keys()) {
      const path = this.keyToPath(pathKey);
      if (path.length > 0 && path[0] === key) {
        return true;
      }
    }
    return false;
  }
}

File: python/collabkit/crdt/map.py
"""
LWW-Map (Last-Writer-Wins Map) CRDT implementation.

A map where each key is independently an LWW-Register.
Supports nested paths for deep object structures.
"""

from __future__ import annotations

from typing import Any, Dict

from .base import CRDT, Operation, StateCRDT, VersionVector


class LWWMap(StateCRDT[Dict[str, Any]]):
    """
    Last-Writer-Wins Map.

    Each key in the map is independently resolved using LWW semantics.
    Supports nested paths like ["users", "123", "name"] for deep updates.

    Example:
        state = LWWMap("node-1")
        state.set(["user", "name"], "Alice")
        state.set(["user", "age"], 30)
        print(state.value())  # {"user": {"name": "Alice", "age": 30}}
    """

    def __init__(self, node_id: str, initial_value: dict[str, Any] | None = None):
        super().__init__(node_id)
        # Store (value, timestamp, node_id) for each path
        self._entries: dict[tuple[str, ...], tuple[Any, float, str]] = {}
        # Track deleted paths
        self._tombstones: dict[tuple[str, ...], tuple[float, str]] = {}

        if initial_value:
            self._init_from_value(initial_value)

    def _init_from_value(self, value: dict[str, Any], path: tuple[str, ...] = ()) -> None:
        """Initialize entries from a nested dictionary."""
        for key, val in value.items():
            current_path = path + (key,)
            if isinstance(val, dict):
                self._init_from_value(val, current_path)
            else:
                self._entries[current_path] = (val, 0.0, self.node_id)

    def set(self, path: list[str] | tuple[str, ...], value: Any) -> Operation:
        """
        Set a value at the given path.

        If value is a dict, it will be flattened into individual path entries.
        Returns the operation that was applied.
        """
        op = Operation.create(
            node_id=self.node_id,
            path=path,
            op_type="set",
            value=value,
        )
        self.apply(op)
        return op

    def delete(self, path: list[str] | tuple[str, ...]) -> Operation:
        """
        Delete a value at the given path.

        Returns the operation that was applied.
        """
        op = Operation.create(
            node_id=self.node_id,
            path=path,
            op_type="delete",
        )
        self.apply(op)
        return op

    def get(self, path: list[str] | tuple[str, ...]) -> Any:
        """Get a value at the given path."""
        path_tuple = tuple(path) if isinstance(path, list) else path

        # Check if path is deleted
        if path_tuple in self._tombstones:
            tombstone = self._tombstones[path_tuple]
            entry = self._entries.get(path_tuple)
            if entry is None or tombstone[0] > entry[1]:
                return None

        entry = self._entries.get(path_tuple)
        if entry:
            return entry[0]

        # Check for nested values
        result = self._get_nested(path_tuple)
        return result

    def _get_nested(self, path: tuple[str, ...]) -> dict[str, Any] | None:
        """Get all values nested under a path as a dictionary."""
        result: dict[str, Any] = {}
        path_len = len(path)

        for entry_path, (value, ts, _) in self._entries.items():
            if len(entry_path) > path_len and entry_path[:path_len] == path:
                # Check tombstone
                tombstone = self._tombstones.get(entry_path)
                if tombstone and tombstone[0] > ts:
                    continue

                # Build nested structure
                remaining = entry_path[path_len:]
                current = result
                for key in remaining[:-1]:
                    if key not in current:
                        current[key] = {}
                    current = current[key]
                current[remaining[-1]] = value

        return result if result else None

    def apply(self, op: Operation) -> bool:
        """Apply a set or delete operation."""
        if self._has_seen(op):
            return False

        path = op.path

        if op.op_type == "set":
            self._apply_set(path, op.value, op.timestamp, op.node_id)
        elif op.op_type == "delete":
            self._apply_delete(path, op.timestamp, op.node_id)
        else:
            raise ValueError(f"LWWMap supports 'set' and 'delete' operations, got '{op.op_type}'")

        self._record_operation(op)
        return True

    def _apply_set(self, path: tuple[str, ...], value: Any, timestamp: float, node_id: str) -> None:
        """Apply a set operation at a path."""
        if isinstance(value, dict):
            # Flatten nested dict into individual entries
            self._flatten_set(path, value, timestamp, node_id)
        else:
            existing = self._entries.get(path)
            if existing is None or self._is_newer(timestamp, node_id, existing[1], existing[2]):
                self._entries[path] = (value, timestamp, node_id)

    def _flatten_set(
        self, path: tuple[str, ...], value: dict[str, Any], timestamp: float, node_id: str
    ) -> None:
        """Flatten a nested dict into individual path entries."""
        for key, val in value.items():
            current_path = path + (key,)
            if isinstance(val, dict):
                self._flatten_set(current_path, val, timestamp, node_id)
            else:
                existing = self._entries.get(current_path)
                if existing is None or self._is_newer(timestamp, node_id, existing[1], existing[2]):
                    self._entries[current_path] = (val, timestamp, node_id)

    def _apply_delete(self, path: tuple[str, ...], timestamp: float, node_id: str) -> None:
        """Apply a delete operation at a path."""
        existing = self._tombstones.get(path)
        if existing is None or self._is_newer(timestamp, node_id, existing[0], existing[1]):
            self._tombstones[path] = (timestamp, node_id)

    def _is_newer(
        self, ts1: float, node1: str, ts2: float, node2: str
    ) -> bool:
        """Check if (ts1, node1) is newer than (ts2, node2)."""
        if ts1 != ts2:
            return ts1 > ts2
        return node1 > node2

    def merge(self, other: "LWWMap") -> None:
        """Merge another map into this one."""
        for op in other.all_operations():
            self.apply(op)

    def value(self) -> dict[str, Any]:
        """Get the current value as a nested dictionary."""
        result: dict[str, Any] = {}

        for path, (value, ts, _) in self._entries.items():
            # Check tombstone
            tombstone = self._tombstones.get(path)
            if tombstone and tombstone[0] > ts:
                continue

            # Build nested structure
            current = result
            for key in path[:-1]:
                if key not in current:
                    current[key] = {}
                elif not isinstance(current[key], dict):
                    # Conflict: path has both a value and nested values
                    # Nested values take precedence
                    current[key] = {}
                current = current[key]

            if path:  # Don't set if path is empty
                current[path[-1]] = value

        return result

    def state(self) -> dict[str, Any]:
        """Get the full state for transmission."""
        return {
            "entries": {
                ".".join(path): {"value": val, "timestamp": ts, "node_id": nid}
                for path, (val, ts, nid) in self._entries.items()
            },
            "tombstones": {
                ".".join(path): {"timestamp": ts, "node_id": nid}
                for path, (ts, nid) in self._tombstones.items()
            },
            "operations": [op.to_dict() for op in self._operations],
        }

    @classmethod
    def from_state(cls, node_id: str, state: dict[str, Any]) -> "LWWMap":
        """Reconstruct map from transmitted state."""
        lww_map = cls(node_id)

        for path_str, entry in state.get("entries", {}).items():
            path = tuple(path_str.split(".")) if path_str else ()
            lww_map._entries[path] = (
                entry["value"],
                entry["timestamp"],
                entry["node_id"],
            )

        for path_str, tombstone in state.get("tombstones", {}).items():
            path = tuple(path_str.split(".")) if path_str else ()
            lww_map._tombstones[path] = (
                tombstone["timestamp"],
                tombstone["node_id"],
            )

        lww_map._operations = [
            Operation.from_dict(op) for op in state.get("operations", [])
        ]

        return lww_map

    def keys(self) -> list[str]:
        """Get top-level keys."""
        seen = set()
        for path in self._entries.keys():
            if path:
                seen.add(path[0])
        return list(seen)

    def __contains__(self, key: str) -> bool:
        """Check if a top-level key exists."""
        return any(path and path[0] == key for path in self._entries.keys())

    def __getitem__(self, key: str) -> Any:
        """Get a value by key (top-level only)."""
        return self.get((key,))

    def __setitem__(self, key: str, value: Any) -> None:
        """Set a value by key (top-level only)."""
        self.set((key,), value)
